File: scripts/field-cad/elements.py
#!/usr/bin/env python
"""
scripts/field-cad/elements.py — BIOBUZZ SCORING-ELEMENT CAD import (the holed POLLEN/NECTAR).

`convert.py` is the FIELD importer and it deliberately DROPS the loose scoring elements
(`RE_ELEMENT`): they are not field structure, and BIOBUZZ spawns them from `config.ts`. But
they ride along in the same STEP assembly as real perforated solids, and the 3D renderer used
to draw them as smooth `THREE.SphereGeometry` spheres. This script is the other half: it pulls
ONE `am-5851: Pollen` and ONE `am-5852: * Nectar` DEFINITION shape out of the same pinned STEP,
measures it, tessellates it, and writes two STL for `elements.mjs` to assemble into
`public/models/biobuzz/elements.glb`.

Run it through `scripts/field-cad/elements.mjs` (`npm run element-cad`), never by hand — that
driver owns the sha-pinned source, the venv and the glTF chain.

── WHAT THE CAD ACTUALLY IS (measured, not assumed) ────────────────────────────────────────
Each element is ONE closed solid with 29 faces: an OUTER sphere, an INNER sphere, and 26
cylindrical BORES drilled on radial axes. Nothing is approximated here — the sphere radii and
the bore radius are exact analytic values read off the B-rep, and this script asserts the
pattern it found (26 bores in a 1/4/8/8/4/1 latitude ring stack) rather than trusting it.

── WHY THE ORIENTATION OF EVERY FACE IS APPLIED, AND `convert.py` DOES NOT ─────────────────
⚠️ OCC hands back a triangulation in the face's OWN parametric winding. For a face whose
`Orientation()` is `TopAbs_REVERSED` that winding is BACKWARDS with respect to the solid, so a
mesh built without flipping those triangles is inconsistently wound — which is exactly the
class of bug `renderFieldGlb.ts`'s `fixGroundBeamWinding` exists to paper over on the field
asset. It matters far more here: a holed ball is SEEN THROUGH ITS OWN HOLES, so its inner
sphere and its 26 bore walls have to render from the inside with ordinary back-face culling.
Flipping on orientation makes every triangle face AWAY FROM THE MATERIAL, which is the one
convention under which that is true, and `verify_winding` below proves it on the outer shell
(every outer-sphere triangle's normal must point away from the centre) before anything is
written.

── FRAME ────────────────────────────────────────────────────────────────────────────────────
The CAD's own pole axis is local +Y. Output is INCHES, centred on the fitted outer-sphere
centre, with the pole on +Z (`sim = (x, -z, y)`, a +90° rotation about X — determinant +1, so
the winding above survives the remap; a mirror would invert every triangle).
"""
from __future__ import annotations

import argparse
import json
import math
import struct
import sys
import time
from collections import defaultdict
from pathlib import Path

from OCP.BRep import BRep_Tool
from OCP.BRepAdaptor import BRepAdaptor_Surface
from OCP.BRepMesh import BRepMesh_IncrementalMesh
from OCP.BRepTools import BRepTools
from OCP.GeomAbs import GeomAbs_SurfaceType
from OCP.STEPCAFControl import STEPCAFControl_Reader
from OCP.TCollection import TCollection_ExtendedString
from OCP.TDataStd import TDataStd_Name
from OCP.TDF import TDF_Label, TDF_LabelSequence
from OCP.TDocStd import TDocStd_Document
from OCP.TopAbs import TopAbs_FACE, TopAbs_REVERSED
from OCP.TopExp import TopExp_Explorer
from OCP.TopLoc import TopLoc_Location
from OCP.TopoDS import TopoDS
from OCP.XCAFApp import XCAFApp_Application
from OCP.XCAFDoc import XCAFDoc_DocumentTool

MM_PER_IN = 25.4

# The two SKUs, and the name each is matched by. One instance of each is enough: every Pollen in
# the assembly is the same part definition, and the two Nectar colours differ only in their
# STYLED_ITEM colour, which this renderer sets per instance anyway (`renderElements.ts`'s
# `NECTAR_COLORS` — and see that file on why blue is NOT the CAD's `plastic#0000ff`).
KINDS = (
    ("pollen", "am-5851", "pollen"),
    ("nectar", "am-5852", "nectar"),
)

# What `src/games/biobuzz/config.ts` says, repeated here the way `convert.py` repeats its own
# six anchors: this is a standalone Python CLI outside the TS build, and a drift between the two
# is exactly what the measurements JSON exists to report. NOT an input to anything — the CAD is
# authoritative for dimensions (owner ruling 2026-09-18) and a mismatch is REPORTED, never
# silently absorbed, because the radius is also a SIM number and changing it is the owner's call.
CONFIG_R_IN = {"pollen": 1.4, "nectar": 1.8}


def label_name(label: TDF_Label) -> str:
    attr = TDataStd_Name()
    if label.FindAttribute(TDataStd_Name.GetID_s(), attr):
        return attr.Get().ToExtString()
    return "<unnamed>"


def find_definitions(step_path: str) -> dict[str, tuple[str, object]]:
    """The DEFINITION (part-local, unplaced) shape of each element kind, by name."""
    app = XCAFApp_Application.GetApplication_s()
    doc = TDocStd_Document(TCollection_ExtendedString("mdtv-xcaf"))
    app.NewDocument(TCollection_ExtendedString("MDTV-XCAF"), doc)
    reader = STEPCAFControl_Reader()
    reader.SetColorMode(True)
    reader.SetNameMode(True)

    t0 = time.time()
    status = reader.ReadFile(step_path)
    print(f"[elements] ReadFile status={status} ({time.time() - t0:.1f}s)", file=sys.stderr)
    if int(status) != 1:
        raise RuntimeError(f"STEP ReadFile failed: {status}")
    t0 = time.time()
    if not reader.Transfer(doc):
        raise RuntimeError("STEP Transfer failed")
    print(f"[elements] Transfer ok ({time.time() - t0:.1f}s)", file=sys.stderr)

    shape_tool = XCAFDoc_DocumentTool.ShapeTool_s(doc.Main())
    free = TDF_LabelSequence()
    shape_tool.GetFreeShapes(free)
    found: dict[str, tuple[str, object]] = {}

    def walk(label: TDF_Label) -> None:
        name = label_name(label)
        if shape_tool.IsReference_s(label):
            ref = TDF_Label()
            if not shape_tool.GetReferredShape_s(label, ref):
                return
            if shape_tool.IsAssembly_s(ref):
                ch = TDF_LabelSequence()
                shape_tool.GetComponents_s(ref, ch)
                for i in range(ch.Length()):
                    walk(ch.Value(i + 1))
                return
            low = name.lower()
            for kind, sku, word in KINDS:
                if kind in found:
                    continue
                if sku.lower() in low or word in low:
                    # the REFERRED label's shape: the part in its own frame, with no assembly
                    # placement composed in. A staged element sits wherever the field designer
                    # dropped it; the mesh this pipeline ships is centred on the origin.
                    found[kind] = (name, shape_tool.GetShape_s(ref))
            return
        if shape_tool.IsAssembly_s(label):
            ch = TDF_LabelSequence()
            shape_tool.GetComponents_s(label, ch)
            for i in range(ch.Length()):
                walk(ch.Value(i + 1))

    for i in range(free.Length()):
        walk(free.Value(i + 1))
    missing = [k for k, _, _ in KINDS if k not in found]
    if missing:
        raise RuntimeError(f"STEP carries no definition for: {', '.join(missing)}")
    return found


# ─────────────────────────────────────────────────────────────────── 1. ANALYTIC MEASUREMENT ──


def measure(shape) -> dict:
    """Read the solid's own B-rep: two sphere radii, the bore radius, the bore axes.

    Everything here comes off the ANALYTIC surface, not off a triangulation, so it is the CAD's
    exact number rather than a tessellation artefact.
    """
    spheres: list[tuple[float, tuple[float, float, float]]] = []
    bores: list[tuple[float, tuple[float, float, float]]] = []
    other = 0
    exp = TopExp_Explorer(shape, TopAbs_FACE)
    while exp.More():
        face = TopoDS.Face_s(exp.Current())
        ad = BRepAdaptor_Surface(face)
        t = ad.GetType()
        if t == GeomAbs_SurfaceType.GeomAbs_Sphere:
            s = ad.Sphere()
            c = s.Location()
            spheres.append((s.Radius(), (c.X(), c.Y(), c.Z())))
        elif t == GeomAbs_SurfaceType.GeomAbs_Cylinder:
            c = ad.Cylinder()
            d = c.Axis().Direction()
            bores.append((c.Radius(), (d.X(), d.Y(), d.Z())))
        else:
            other += 1
        exp.Next()

    if other:
        raise RuntimeError(f"unexpected non-sphere/cylinder faces: {other}")
    radii = sorted({round(r, 6) for r, _ in spheres}, reverse=True)
    if len(radii) != 2:
        raise RuntimeError(f"expected exactly two distinct sphere radii, got {radii}")
    outer_r, inner_r = radii
    centres = [c for r, c in spheres if abs(r - outer_r) < 1e-6]
    cx = sum(c[0] for c in centres) / len(centres)
    cy = sum(c[1] for c in centres) / len(centres)
    cz = sum(c[2] for c in centres) / len(centres)

    bore_radii = sorted({round(r, 6) for r, _ in bores})
    if len(bore_radii) != 1:
        raise RuntimeError(f"expected one bore radius, got {bore_radii}")

    # ── THE PATTERN, DERIVED AND THEN ASSERTED ─────────────────────────────────────────────
    # A bore axis is signed, and the two ends of one drilled hole are ONE cylinder face here
    # (26 faces, 26 bores), so the latitude bands are read straight off the pole component.
    # The pole is the CAD's local +Y.
    bands: dict[float, int] = defaultdict(int)
    for _, d in bores:
        bands[round(d[1], 3)] += 1
    band_list = sorted(bands.items(), key=lambda kv: -kv[0])
    return {
        "outerRadiusMm": outer_r,
        "innerRadiusMm": inner_r,
        "wallMm": outer_r - inner_r,
        "boreRadiusMm": bore_radii[0],
        "boreCount": len(bores),
        "centreMm": (cx, cy, cz),
        "bands": [{"sinLat": s, "count": n, "latDeg": math.degrees(math.asin(max(-1.0, min(1.0, s))))} for s, n in band_list],
        "boreAxes": [(round(d[0], 6), round(d[1], 6), round(d[2], 6)) for _, d in bores],
    }


# ───────────────────────────────────────────────────────────────────────── 2. TESSELLATION ──


def mesh(shape, lin_mm: float, ang_rad: float, centre_mm, scale_in: float):
    """(positions_in, triangles) in the SIM frame, every triangle wound AWAY FROM THE MATERIAL.

    ⚠️ `BRepTools.Clean_s` first — `BRepMesh_IncrementalMesh` caches on the shape's own TShape
    and silently no-ops at a second deflection (the same trap `convert.py`'s header records).
    """
    BRepTools.Clean_s(shape)
    BRepMesh_IncrementalMesh(shape, lin_mm, False, ang_rad, True)
    positions: list[tuple[float, float, float]] = []
    tris: list[tuple[int, int, int]] = []
    dedup: dict[tuple[float, float, float], int] = {}
    cx, cy, cz = centre_mm

    exp = TopExp_Explorer(shape, TopAbs_FACE)
    while exp.More():
        face = TopoDS.Face_s(exp.Current())
        loc = TopLoc_Location()
        tri = BRep_Tool.Triangulation_s(face, loc)
        if tri is not None:
            trsf = loc.Transformation()
            # SEE THIS FILE'S HEADER. A reversed face's triangulation is wound backwards with
            # respect to the solid; without this flip the inner sphere and the 26 bore walls
            # are invisible from inside the ball, which is the only place they are ever seen.
            reversed_face = face.Orientation() == TopAbs_REVERSED
            local: dict[int, int] = {}
            for i in range(1, tri.NbNodes() + 1):
                p = tri.Node(i)
                p.Transform(trsf)
                # recentre in mm, scale to inches, then (x, -z, y): pole +Y -> sim +Z, a
                # rotation (det +1), so the winding established above is preserved.
                x = (p.X() - cx) * scale_in
                y = (p.Y() - cy) * scale_in
                z = (p.Z() - cz) * scale_in
                key = (round(x, 5), round(-z, 5), round(y, 5))
                gi = dedup.get(key)
                if gi is None:
                    gi = len(positions)
                    positions.append(key)
                    dedup[key] = gi
                local[i] = gi
            for t in range(1, tri.NbTriangles() + 1):
                n1, n2, n3 = tri.Triangle(t).Get()
                a, b, c = local[n1], local[n2], local[n3]
                if a == b or b == c or a == c:
                    continue  # a degenerate collapsed by the dedup above
                tris.append((a, c, b) if reversed_face else (a, b, c))
        exp.Next()
    return positions, tris


def verify_winding(positions, tris, outer_r_in: float) -> dict:
    """Every triangle whose three vertices sit ON the outer sphere must face OUTWARD.

    This is the single proof that the orientation flip above is right way round. It is checked
    on the outer shell because that is the one surface whose correct normal direction is known
    without any topology: away from the centre. If the flip were inverted, this count would be
    the whole outer shell rather than zero.
    """
    on_outer = 0
    bad = 0
    eps = outer_r_in * 0.02
    for a, b, c in tris:
        pa, pb, pc = positions[a], positions[b], positions[c]
        if not all(abs(math.dist(p, (0, 0, 0)) - outer_r_in) < eps for p in (pa, pb, pc)):
            continue
        on_outer += 1
        ux, uy, uz = (pb[0] - pa[0], pb[1] - pa[1], pb[2] - pa[2])
        vx, vy, vz = (pc[0] - pa[0], pc[1] - pa[1], pc[2] - pa[2])
        nx, ny, nz = (uy * vz - uz * vy, uz * vx - ux * vz, ux * vy - uy * vx)
        mx = (pa[0] + pb[0] + pc[0]) / 3
        my = (pa[1] + pb[1] + pc[1]) / 3
        mz = (pa[2] + pb[2] + pc[2]) / 3
        if nx * mx + ny * my + nz * mz <= 0:
            bad += 1
    return {"outerTris": on_outer, "inwardFacing": bad}


def write_binary_stl(path: Path, positions, tris) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    with open(path, "wb") as f:
        f.write(b"\x00" * 80)
        f.write(struct.pack("<I", len(tris)))
        for a, b, c in tris:
            ax, ay, az = positions[a]
            bx, by, bz = positions[b]
            cx, cy, cz = positions[c]
            ux, uy, uz = bx - ax, by - ay, bz - az
            vx, vy, vz = cx - ax, cy - ay, cz - az
            nx = uy * vz - uz * vy
            ny = uz * vx - ux * vz
            nz = ux * vy - uy * vx
            nl = math.sqrt(nx * nx + ny * ny + nz * nz) or 1.0
            f.write(struct.pack("<3f", nx / nl, ny / nl, nz / nl))
            f.write(struct.pack("<3f", ax, ay, az))
            f.write(struct.pack("<3f", bx, by, bz))
            f.write(struct.pack("<3f", cx, cy, cz))
            f.write(struct.pack("<H", 0))


def main() -> None:
    ap = argparse.ArgumentParser()
    ap.add_argument("step")
    ap.add_argument("--out", required=True, help="directory for the STL + measurements JSON")
    # THE ONE TUNING KNOB, and it is a LENGTH so it means the same thing on both elements: the
    # chord tolerance in inches. `elements.mjs` owns the shipped value and the triangle budget
    # it was picked against; the sweep that picked it is in this session's report.
    ap.add_argument("--deflection", type=float, default=0.012, help="linear chord deflection, INCHES")
    ap.add_argument("--angular", type=float, default=0.55, help="angular deflection, radians")
    ap.add_argument("--sweep", action="store_true", help="report triangle counts over a range and write nothing")
    args = ap.parse_args()

    found = find_definitions(args.step)
    out = Path(args.out)
    report: dict[str, dict] = {}

    for kind, _sku, _word in KINDS:
        name, shape = found[kind]
        m = measure(shape)
        outer_in = m["outerRadiusMm"] / MM_PER_IN
        centre = m["centreMm"]
        print(f"\n=== {kind.upper()}  '{name}' ===")
        print(f"  outer   {m['outerRadiusMm']:.4f} mm = {outer_in:.5f} in  (config {CONFIG_R_IN[kind]})")
        print(f"  inner   {m['innerRadiusMm']:.4f} mm = {m['innerRadiusMm'] / MM_PER_IN:.5f} in")
        print(f"  wall    {m['wallMm']:.4f} mm = {m['wallMm'] / MM_PER_IN:.5f} in")
        print(f"  bores   {m['boreCount']} x r {m['boreRadiusMm']:.4f} mm = d {2 * m['boreRadiusMm'] / MM_PER_IN:.5f} in")
        print(f"  bands   {[(b['count'], round(b['latDeg'], 2)) for b in m['bands']]}")

        if args.sweep:
            for d in (0.03, 0.02, 0.015, 0.012, 0.01, 0.008, 0.006, 0.004):
                for ang in (0.9, 0.7, 0.55, 0.4):
                    pos, tris = mesh(shape, d * MM_PER_IN, ang, centre, 1.0 / MM_PER_IN)
                    w = verify_winding(pos, tris, outer_in)
                    print(f"    lin={d:<6} ang={ang:<5} verts={len(pos):<6} tris={len(tris):<6} badWinding={w['inwardFacing']}")
            continue

        pos, tris = mesh(shape, args.deflection * MM_PER_IN, args.angular, centre, 1.0 / MM_PER_IN)
        w = verify_winding(pos, tris, outer_in)
        if w["inwardFacing"]:
            raise RuntimeError(f"{kind}: {w['inwardFacing']} outer-shell triangles face inward — the orientation flip is wrong")
        rmax = max(math.dist(p, (0, 0, 0)) for p in pos)
        write_binary_stl(out / f"{kind}.stl", pos, tris)
        print(f"  mesh    {len(pos)} verts, {len(tris)} tris, outer-shell winding clean ({w['outerTris']} tris checked)")
        report[kind] = {
            "part": name,
            "outerRadiusIn": round(outer_in, 6),
            "innerRadiusIn": round(m["innerRadiusMm"] / MM_PER_IN, 6),
            "wallIn": round(m["wallMm"] / MM_PER_IN, 6),
            "boreRadiusIn": round(m["boreRadiusMm"] / MM_PER_IN, 6),
            "boreDiameterIn": round(2 * m["boreRadiusMm"] / MM_PER_IN, 6),
            "boreCount": m["boreCount"],
            "bands": [{"count": b["count"], "latitudeDeg": round(b["latDeg"], 4)} for b in m["bands"]],
            "configRadiusIn": CONFIG_R_IN[kind],
            "configDeltaIn": round(outer_in - CONFIG_R_IN[kind], 6),
            "meshVerts": len(pos),
            "meshTris": len(tris),
            "meshMaxRadiusIn": round(rmax, 6),
            "deflectionIn": args.deflection,
            "angularDeflection": args.angular,
        }

    if args.sweep:
        return
    out.mkdir(parents=True, exist_ok=True)
    (out / "elements-measurements.json").write_text(json.dumps(report, indent=2), encoding="utf8")
    print(f"\n[elements] wrote {out / 'elements-measurements.json'}")


if __name__ == "__main__":
    main()

File: scripts/field-cad/elements.mjs
#!/usr/bin/env node
// scripts/field-cad/elements.mjs — BIOBUZZ SCORING-ELEMENT CAD pipeline (`npm run element-cad`).
//
// The sibling of `scripts/field-cad.mjs`, for the one thing that file deliberately throws away.
// `convert.py`'s `RE_ELEMENT` drops the staged POLLEN and NECTAR because they are not field
// STRUCTURE — but they are in the same STEP assembly as real perforated solids, and the 3D
// renderer drew them as smooth spheres. This driver:
//
//   (a) resolves the SAME sha-pinned STEP zip from the SAME out-of-repo cache (nothing is
//       re-downloaded when the cached zip's hash still matches `source.mjs`);
//   (b) runs `elements.py` with the cache venv's Python to measure and tessellate one Pollen
//       and one Nectar definition into inch-scale, origin-centred STL;
//   (c) assembles the two into ONE glTF with two named meshes, and runs gltf-transform's
//       weld -> meshopt chain, exactly like the field's two LODs;
//   (d) prints the measured-vs-budget table and FAILS on a triangle or size overrun.
//
// Output: `public/models/biobuzz/elements.glb` + `elements-measurements.json`. Both ship under
// the same licence decision as the field files — see that directory's README.
//
// ── NO `simplify` PASS, DELIBERATELY ────────────────────────────────────────────────────────
// A decimator's error metric is a distance, and the feature that matters here is a 0.44-in hole
// in a 2.8-in ball: any tolerance loose enough to save triangles closes hole rims first, and the
// holes are the entire point of the asset. The triangle count is controlled at the SOURCE
// instead, by the chord deflection below, which is a tolerance on the true analytic surface
// rather than on an already-tessellated one.
import { existsSync, mkdirSync, readFileSync, writeFileSync, statSync } from 'node:fs';
import { execFileSync } from 'node:child_process';
import path from 'node:path';
import zlib from 'node:zlib';
import crypto from 'node:crypto';
import { fileURLToPath } from 'node:url';

import { SOURCE } from './source.mjs';

const CACHE = 'C:/Users/geniu/AppData/Local/dsim/field-cad';
const VENV_PY = path.join(CACHE, 'venv', 'Scripts', 'python.exe');
const NPMTOOLS = path.join(CACHE, 'npmtools');
const GLTF_TRANSFORM = path.join(NPMTOOLS, 'node_modules', '@gltf-transform', 'cli', 'bin', 'cli.js');

const REPO_ROOT = path.resolve(path.dirname(fileURLToPath(import.meta.url)), '..', '..');
const PUBLIC_DIR = path.join(REPO_ROOT, 'public', 'models', 'biobuzz');
const WORK = path.join(CACHE, 'elements');

// ── THE TESSELLATION, AND THE BUDGET IT WAS PICKED AGAINST ──────────────────────────────────
// A chord deflection in INCHES, so it means the same thing on both elements. Measured sweep
// (`elements.py --sweep`, tris pollen/nectar): 0.03 -> 2286/2542, 0.02 -> 2684/2968,
// 0.015 -> 3034/3610, 0.01 -> 3940/4786, 0.004 -> 8174/9962. 0.03 in is ~12 deg of arc on a
// 1.4-in ball (a 30-segment great circle) and ~12 segments around a 0.44-in bore, which is past
// the point where either reads as faceted at any camera distance this game has; everything
// below it buys silhouette nobody can see at the cost of triangles a full field pays 100 times.
const DEFLECTION_IN = 0.03;
const ANGULAR = 0.9;

// PER BALL, and the reason it is a ratchet: a match has up to ~100 elements on screen across two
// `InstancedMesh`es, and each one is drawn again in the sun's shadow pass on High/Ultra.
const TRI_BUDGET = 3000;
const BUDGET_BROTLI = 60 * 1024;

function log(msg) {
  console.log(`[element-cad] ${msg}`);
}

function sha256File(p) {
  return crypto.createHash('sha256').update(readFileSync(p)).digest('hex');
}

/** the same pinned zip `scripts/field-cad.mjs` resolves, in the same cache, by the same rule. */
function ensureSource() {
  mkdirSync(CACHE, { recursive: true });
  const zipPath = path.join(CACHE, 'field-cad-step.zip');
  const extractedDir = path.join(CACHE, 'extracted');
  const stepPath = path.join(extractedDir, 'field-cad-step.step');

  let needDownload = true;
  if (existsSync(zipPath)) {
    const actual = sha256File(zipPath);
    if (actual === SOURCE.sha256) {
      needDownload = false;
      log(`cached zip sha256 OK (${actual}) — skipping download`);
    } else {
      log(`WARNING: cached zip sha256 ${actual} != pinned ${SOURCE.sha256} — re-downloading`);
    }
  }
  if (needDownload) {
    log(`downloading ${SOURCE.url}`);
    execFileSync('curl', ['-sL', '-o', zipPath, SOURCE.url], { stdio: 'inherit' });
    const actual = sha256File(zipPath);
    if (actual !== SOURCE.sha256) {
      throw new Error(`sha256 mismatch: got ${actual}, pinned SOURCE.sha256 is ${SOURCE.sha256}.`);
    }
  }
  if (!existsSync(stepPath)) {
    mkdirSync(extractedDir, { recursive: true });
    log('extracting zip');
    execFileSync('tar', ['-xf', zipPath, '-C', extractedDir], { stdio: 'inherit' });
  }
  return stepPath;
}

function runExtract(stepPath) {
  if (!existsSync(VENV_PY)) {
    throw new Error(`python venv not found at ${VENV_PY}. See scripts/field-cad.mjs for the one-time setup.`);
  }
  mkdirSync(WORK, { recursive: true });
  log('running elements.py (CadQuery/OCP, headless) ...');
  execFileSync(
    VENV_PY,
    [path.join(REPO_ROOT, 'scripts', 'field-cad', 'elements.py'), stepPath, '--out', WORK, '--deflection', String(DEFLECTION_IN), '--angular', String(ANGULAR)],
    { stdio: 'inherit' },
  );
}

/** binary STL -> { positions: Float32Array, indices: Uint16Array } with vertices welded on an
 * exact position match (the STL was written from an already-deduplicated index buffer, so this
 * recovers that buffer rather than approximating it). */
function readStl(file) {
  const buf = readFileSync(file);
  const tris = buf.readUInt32LE(80);
  const map = new Map();
  const pos = [];
  const idx = [];
  for (let t = 0; t < tris; t++) {
    const base = 84 + t * 50 + 12; // skip the 80-byte header, the count, and this facet's normal
    for (let v = 0; v < 3; v++) {
      const x = buf.readFloatLE(base + v * 12);
      const y = buf.readFloatLE(base + v * 12 + 4);
      const z = buf.readFloatLE(base + v * 12 + 8);
      const key = `${x},${y},${z}`;
      let i = map.get(key);
      if (i === undefined) {
        i = pos.length / 3;
        pos.push(x, y, z);
        map.set(key, i);
      }
      idx.push(i);
    }
  }
  if (pos.length / 3 > 65535) throw new Error(`${file}: ${pos.length / 3} verts exceeds UNSIGNED_SHORT indexing`);
  return { positions: new Float32Array(pos), indices: new Uint16Array(idx), tris };
}

/**
 * Write a minimal two-mesh GLB by hand. It is a JSON chunk plus a binary chunk, and building it
 * here rather than through `@gltf-transform/core` keeps this driver runnable with nothing but
 * Node — the gltf-transform CLI is still used below for weld + meshopt, which is where the real
 * work is.
 *
 * POSITION AND INDICES ONLY, NO NORMALS — the same convention both field GLBs ship under (see
 * `public/models/biobuzz/README.md`). `renderElementsGlb.ts` runs `computeCreasedNormals` at
 * load, which is what gives a smooth sphere AND a hard rim where each bore breaks the surface;
 * baked normals would need a vertex split at every one of those rims for no visual gain.
 */
function writeGlb(outPath, meshes) {
  const json = {
    asset: { version: '2.0', generator: 'dsim scripts/field-cad/elements.mjs' },
    scene: 0,
    scenes: [{ nodes: meshes.map((_, i) => i) }],
    nodes: meshes.map((m, i) => ({ mesh: i, name: m.name })),
    meshes: meshes.map((m, i) => ({ name: m.name, primitives: [{ attributes: { POSITION: i * 2 }, indices: i * 2 + 1 }] })),
    accessors: [],
    bufferViews: [],
    buffers: [],
  };
  const blobs = [];
  let offset = 0;
  const push = (data) => {
    // 4-byte alignment: every accessor here is 4- or 2-byte and glTF requires the view's offset
    // to be a multiple of the component size.
    const pad = (4 - (offset % 4)) % 4;
    if (pad) {
      blobs.push(Buffer.alloc(pad));
      offset += pad;
    }
    const b = Buffer.from(data.buffer, data.byteOffset, data.byteLength);
    blobs.push(b);
    const view = { buffer: 0, byteOffset: offset, byteLength: b.length };
    offset += b.length;
    return view;
  };

  for (const m of meshes) {
    const { positions, indices } = m;
    let minX = Infinity;
    let minY = Infinity;
    let minZ = Infinity;
    let maxX = -Infinity;
    let maxY = -Infinity;
    let maxZ = -Infinity;
    for (let i = 0; i < positions.length; i += 3) {
      minX = Math.min(minX, positions[i]);
      maxX = Math.max(maxX, positions[i]);
      minY = Math.min(minY, positions[i + 1]);
      maxY = Math.max(maxY, positions[i + 1]);
      minZ = Math.min(minZ, positions[i + 2]);
      maxZ = Math.max(maxZ, positions[i + 2]);
    }
    json.bufferViews.push({ ...push(positions), target: 34962 });
    json.accessors.push({
      bufferView: json.bufferViews.length - 1,
      componentType: 5126,
      count: positions.length / 3,
      type: 'VEC3',
      min: [minX, minY, minZ],
      max: [maxX, maxY, maxZ],
    });
    json.bufferViews.push({ ...push(indices), target: 34963 });
    json.accessors.push({ bufferView: json.bufferViews.length - 1, componentType: 5123, count: indices.length, type: 'SCALAR' });
  }

  let bin = Buffer.concat(blobs);
  if (bin.length % 4) bin = Buffer.concat([bin, Buffer.alloc(4 - (bin.length % 4))]);
  json.buffers.push({ byteLength: bin.length });

  let jsonBuf = Buffer.from(JSON.stringify(json), 'utf8');
  if (jsonBuf.length % 4) jsonBuf = Buffer.concat([jsonBuf, Buffer.alloc(4 - (jsonBuf.length % 4), 0x20)]);

  const header = Buffer.alloc(12);
  header.writeUInt32LE(0x46546c67, 0); // 'glTF'
  header.writeUInt32LE(2, 4);
  header.writeUInt32LE(12 + 8 + jsonBuf.length + 8 + bin.length, 8);
  const jsonHeader = Buffer.alloc(8);
  jsonHeader.writeUInt32LE(jsonBuf.length, 0);
  jsonHeader.writeUInt32LE(0x4e4f534a, 4); // 'JSON'
  const binHeader = Buffer.alloc(8);
  binHeader.writeUInt32LE(bin.length, 0);
  binHeader.writeUInt32LE(0x004e4942, 4); // 'BIN'
  writeFileSync(outPath, Buffer.concat([header, jsonHeader, jsonBuf, binHeader, bin]));
}

function gltfTransform(cmd, args) {
  execFileSync('node', [GLTF_TRANSFORM, cmd, ...args], { cwd: NPMTOOLS, stdio: 'inherit' });
}

function main() {
  log(`source: ${SOURCE.url} (${SOURCE.version}, ${SOURCE.versionDate})`);
  const stepPath = ensureSource();
  runExtract(stepPath);

  const measurements = JSON.parse(readFileSync(path.join(WORK, 'elements-measurements.json'), 'utf8'));
  const kinds = ['pollen', 'nectar'];
  const meshes = kinds.map((k) => ({ name: k, ...readStl(path.join(WORK, `${k}.stl`)) }));

  const raw = path.join(WORK, 'elements-raw.glb');
  const welded = path.join(WORK, 'elements-welded.glb');
  const out = path.join(PUBLIC_DIR, 'elements.glb');
  mkdirSync(PUBLIC_DIR, { recursive: true });
  writeGlb(raw, meshes);
  gltfTransform('weld', [raw, welded]);
  gltfTransform('meshopt', [welded, out]);

  // the measurements file ships too: `renderElementsGlb.ts` never reads it (the mesh IS the
  // truth at runtime), but the smoke lane and a future session both need the CAD's own numbers
  // next to the asset they describe.
  const measOut = path.join(PUBLIC_DIR, 'elements-measurements.json');
  writeFileSync(measOut, `${JSON.stringify(measurements, null, 2)}\n`);

  const rawBytes = statSync(out).size;
  const brotli = zlib.brotliCompressSync(readFileSync(out)).length;

  console.log('');
  console.log('size table (measured vs budget):');
  for (const m of meshes) {
    const ok = m.tris <= TRI_BUDGET ? 'OK' : 'OVER';
    console.log(`  ${m.name.padEnd(7)} verts=${(m.positions.length / 3).toString().padStart(5)}  tris=${m.tris.toString().padStart(5)}  budget=${TRI_BUDGET}  ${ok}`);
  }
  console.log(`  elements.glb   raw=${rawBytes.toLocaleString()}  brotli=${brotli.toLocaleString()}  budget(brotli)=${BUDGET_BROTLI.toLocaleString()}  ${brotli <= BUDGET_BROTLI ? 'OK' : 'OVER'}`);
  for (const k of kinds) {
    const m = measurements[k];
    console.log(
      `  ${k.padEnd(7)} CAD r=${m.outerRadiusIn} in (config ${m.configRadiusIn}, delta ${m.configDeltaIn})  wall=${m.wallIn}  ${m.boreCount} bores d=${m.boreDiameterIn}`,
    );
  }
  console.log('');

  let failed = false;
  for (const m of meshes) {
    if (m.tris > TRI_BUDGET) {
      failed = true;
      console.error(`FAIL: ${m.name} is ${m.tris} triangles > budget ${TRI_BUDGET}.`);
    }
  }
  if (brotli > BUDGET_BROTLI) {
    failed = true;
    console.error(`FAIL: elements.glb is ${brotli} brotli bytes > budget ${BUDGET_BROTLI}.`);
  }
  if (failed) process.exit(1);
  log(`wrote ${out} and ${measOut} — ALL BUDGETS OK`);
}

main();
